File: csci3180 asg2/task4_java/Task4Soldier.java
/*
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 2
 * Name : Cheung Kam Fung
 * Student ID : 1155110263
 * Email Addr : 1155110263@link.cuhk.edu.hk
 */

import java.util.HashSet;

public class Task4Soldier extends Soldier{
  private int coins;
  private int defence;

  public Task4Soldier() {
    super();
    this.coins = 0;
    this.defence = 0;
    /* It is possible for the soldier to obtain two identical keys in the game.
       To simplify the operation, we use set here to remove duplicate keys. */
  }

  public void task4AddCoins() { this.coins += 1; } // add Coins

  public void task4BuyElixirs() { // buy Exlixirs from merchant
    super.addElixir();
    this.coins -= 1;
  }

  public boolean loseHealth() {
    this.health -= (10 - this.defence) >= 0 ? 10 - this.defence : 0 ; // value > 0
    return this.health <= 0;
  }

  public void task4BuyDefence() { // but defence from merchant
    this.defence += 5;
    this.coins -= 2;
  }

  public int task4GetCoins() { return this.coins; } // return coins

  public void displayInformation() {
    super.displayInformation();
    System.out.printf("Defence: %d.%n", this.defence);
    System.out.printf("Coins: %d.%n", this.coins);
  }


}

File: csci3180 asg2/task4_python/Task4Soldier.py
"""
/*
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 2
 * Name : Cheung Kam Fung
 * Student ID : 1155110263
 * Email Addr : 1155110263@link.cuhk.edu.hk
 */
 """
from Soldier import Soldier


class Task4Solider(Soldier):
    def __init__(self):
        self.coins = 0
        self.defence = 0
        super().__init__()

    def task4AddCoins(self):
        self.coins += 1

    def task4BuyElixirs(self):
        super().addElixir()
        self.coins -= 1

    def loseHealth(self):
        if self.defence >= 10:
            self.health -= 0 # minimum loseheal is 0
        else:
            self.health -= (10 - self.defence) #minus defence

        return self.health <= 0

    def task4BuyDefence(self):
        self.defence += 5
        self.coins -= 2

    def task4GetCoins(self):
        return self.coins

    def displayInformation(self):
        super().displayInformation()
        print("Defence: ", self.defence, ".", sep='')
        print("Coins: ", self.coins, ".", sep='')
